### ISSUE TITLE
Feature request: Installability as a PyPI package

### ISSUE THREAD

**Issue by csachs** [aristocratos/bpytop#2]
I think it would be easier (a one-liner) to install, if bpytop would be packaged as Python package, and be uploaded to the Python Package Index (PyPI).

**Comment by aristocratos** [comment]
Gonna wait a little while so I know it's properly stable. Also gonna need to write some tests.
But I have planned to set up automated testing and pypi packaging later on.

**Comment by csachs** [comment]
Ah good to hear, I thought due to version 1.0 being released it was of suitable stability to be packaged.
Will wait for a package then (and install from git meanwhile) :)

**Comment by aristocratos** [comment]
Yeah, probably shouldn't have started at 1.0.0, but I wanted the versioning to be higher than that of bashtop to drive people to bpytop, and save me the issue of bug fixing a 5000 line bash script...

**Comment by aristocratos** [comment]
I wrote this in the readme but I'm gonna repeat it here. If anybody is willing to help with writing tests for bpytop that would be very helpful. And would speed up the process of making it suitable for PyPi packaging.

**Comment by mbarkhau** [comment]
I don't know what makes sense to test yet, wouldn't it make sense to do some project setup, to make it easier for contributors to get started?

I usually setup my projects using https://gitlab.com/mbarkhau/bootstrapit
An example of such a project is: https://gitlab.com/mbarkhau/markdown-katex

I did the setup for this project yesterday, which took maybe an hour: https://gitlab.com/mbarkhau/pretty-traceback

I personally would publish to pypi, regardless of test coverage, just so long as you have a big enough disclaimer about the hacky pre-alpha prototype status of the project. This would also prevent anybody from namesquatting `bpytop` on pypi.

**Comment by aristocratos** [comment]
> I don't know what makes sense to test yet

Neither do I, this is the first piece of python code longer than 100 lines I've ever written and I have no experience in setting up testing. I'll get to it eventually, but the reason in asking for help is to speed the process up a bit.

 > This would also prevent anybody from namesquatting bpytop on pypi

Good point.

> the hacky pre-alpha prototype status of the project

I'm not sure what your expectations of the code quality is. This isn't production code with critical systems depending on it but a a hobby project I wrote while learning python during my vacation. I'll keep bug fixing reported errors and refactor code as I learn better methods, but that's about it.

**Comment by mbarkhau** [comment]
> > the hacky pre-alpha prototype status of the project
> I'm not sure what your expectations of the code quality is.

I was just doing hyperbol ;-)

**Comment by bpepple** [comment]
> > I don't know what makes sense to test yet
> 
> Neither do I, this is the first piece of python code longer than 100 lines I've ever written and I have no experience in setting up testing. I'll get to it eventually, but the reason in asking for help is to speed the process up a bit.

I would suggest that it makes sense to split some classes/ functions out into separate file (like for example the config code), and then writing the tests would be much easier. Right now the main file is pretty huge and make it smaller would make maintenance a bit easier. If your interested I can probably help some time this week.

**Comment by aristocratos** [comment]
It might not be a popular opinion but I actually prefer one big file over multiple smaller files, my reasoning being:

1. Modern IDE's usually have some sort of symbol overview, for example vs codes symbols sidebar which makes it really easy jumping around functions and classes in the tree view.

2. Having one file makes it easier for the end user; just drop the .py file in path and run. Although I guess this could be done via a script patching up multiple files into one at install. But as it is right now, anyone could simply download "bpytop.py" and have everything they need.

The line count should likely start to drop a bit when I begin refactoring since I've learned a lot since I started this project.

**Comment by bpepple** [comment]
> It might not be a popular opinion but I actually prefer one big file over multiple smaller files, my reasoning being:

>     2. Having one file makes it easier for the end user; just drop the .py file in path and run. Although I guess this could be done via a script patching up multiple files into one at install. But as it is right now, anyone could simply download "bpytop.py" and have everything they need.

Generally speaking the end user shouldn't need to care if it's one file or many, and there's a reason why setuptools is used fairly universally in python packages.

**Comment by aristocratos** [comment]
> Generally speaking the end user shouldn't need to care if it's one file or many, and there's a reason why setuptools is used fairly universally in python packages.

No, I definitively agree you. But I have no previous experience with setuptools and haven't had time to read through the documentation yet. 

Don't have much time or energy left after work during normal weekdays so most of my focus have been on bug fixing and finishing missing features while testing and packaging have fallen behind. Which I know is bad practice, but I also don't have deadline and it is open source, so my hope is that others will help out if they are interested in the project.

**Comment by bpepple** [comment]
> Don't have much time or energy left after work during normal weekdays so most of my focus have been on bug fixing and finishing missing features while testing and packaging have fallen behind. Which I know is bad practice, but I also don't have deadline and it is open source, so my hope is that others will help out if they are interested in the project.

Definitely understandable, but really you're just kicking the can down the road and adding extra work for yourself later by putting off improving the packaging. And frankly I wouldn't even worry about testing until that is completed, since most likely you're going to need to modify the file layout for the project and refactor the main file (primarily so you have a clean entry point for the console script). At a minimum, I'd create a branch for this for anyone that would want to help in this effort.

**Comment by jotyGill** [comment]
First of all thank you for making this!
I have name squatted 'bpytop' on pypi, give me your monies (haha)
I just secured it so no one else can steal it. when you're ready to publish it I'll transfer the ownership to you :)
You'll have to make some changes in code to work with relative paths instead of hard coded /usr/local/share/themes
Because when installing stuff using pip different OS's install it in different locations, it also depends if users install it using `sudo pyhton3 -m pip install`  vs `python3  -m pip install --user`
Also the standard way to structure a python project is to have the setup.py at the base then project folder with the .py files and data folders. I can send a basic pull to show what I mean. I'll try to help out if I can.
